src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.cs

```diff
@@ -208,6 +217,19 @@ public async Task ProcessRequestsAsync<TContext>(IHttpApplication<TContext> appl
                     // Call UpdateCompletedStreams() prior to frame processing in order to remove any streams that have exceeded their drain timeouts.
                     UpdateCompletedStreams();
 
+                    if (_keepAlive != null)
+                    {
+                        var state = _keepAlive.ProcessKeepAlive(!result.IsCanceled);
+                        if (state == KeepAliveState.SendPing)
+                        {
+                            await _frameWriter.WritePingAsync(Http2PingFrameFlags.NONE, Http2ConnectionKeepAlive.PingPayload);
+                        }
+                        else if (state == KeepAliveState.Timeout)
+                        {
+                            throw new Http2ConnectionErrorException(CoreStrings.Http2ErrorKeepAliveTimeout, Http2ErrorCode.CANCEL);
```

Comment:
Right error code to send with GO_AWAY?

Reply:
No, cancel is for streams. Elsewhere we use Internal_error when aborting the connection.
https://github.com/dotnet/aspnetcore/blob/6ac6337beb576defc77125cdf43ec5c88c998cd7/src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.cs#L152

Reply:
I don't see anything that obviously makes sense among the [predefined HTTP/2 error codes](https://http2.github.io/http2-spec/#ErrorCodes). Arguably this doesn't matter much since a ping timeout implies the client isn't properly processing frames, but it's possible that the client will see the GOAWAY frame eventually. Do you know what node does?

Reply:
I'm OK with INTERNAL_ERROR.

Reply:
Go sends NO_ERROR

Reply:
I think if there is a ping timeout then golang/Netty just close the TCP connection. That would make sense if you assume the client is no longer there.

Reply:
Yeah. GoAway is a best effort thing regardless.

Reply:
Ok. I don't like INTERNAL_ERROR.

> INTERNAL_ERROR (0x2):  The endpoint encountered an unexpected
>       internal error.

NO_ERROR makes more sense to me.

Reply:
I don't like NO_ERROR because the spec says that means "Graceful shutdown" which implies that we'll keep accepting frames for active streams, but no new streams should be created. We're abruptly stopping the entire frame processing loop whether or not there are active streams, right? That doesn't seem graceful.